### DISCUSSION TITLE
Some errors occured in MDBC method

### DISCUSSION THREAD

**Discussion by chenxin202** [DualSPHysics/DualSPHysics#285] [DUALSPHYSICS_V5.2]
Dear Developer team
   When I tried to use MDBC method to simulate a floating structure including a rectangular floating body and several cylinders. I found some numerical gaps occurred near fluid-rectangular floatings interface. But when I removed the cylinders, I ran the numerical case only including the rectangular structure using the same configuration. The gap will disappear. Could you give me some advice to avoid these problems? 
The attached files are the snapshot of the numerical gap and the related Def file.
![first 1](https://github.com/user-attachments/assets/0543248a-59f9-4661-8bf9-3267933219d5)

<?xml version="1.0" encoding="UTF-8" ?>
<case app="GenCase v5.0.278c (30-10-2023)" date="09-12-2024 00:27:34">
    <casedef>
        <constantsdef>
            
            <gravity x="0" y="0" z="-9.81" comment="Gravitational acceleration" units_comment="m/s^2" />
            <rhop0 value="1000" comment="Reference density of the fluid" units_comment="kg/m^3" />
            <rhopgradient value="3" comment="Initial density gradient 1:Rhop0, 2:Water column, 3:Max. water height (default=2)" />
            <hswl value="0" auto="true" comment="Maximum still water level to calculate speedofsound using coefsound" units_comment="metres (m)" />
            <gamma value="7" comment="Polytropic constant for water used in the state equation" />
            <speedsystem value="0" auto="true" comment="Maximum system speed (by default the dam-break propagation is used)" />
            <coefsound value="20" comment="Coefficient to multiply speedsystem" />
            <speedsound value="0" auto="true" comment="Speed of sound to use in the simulation (by default speedofsound=coefsound*speedsystem)" />
            <coefh value="1.5" comment="Coefficient to calculate the smoothing length (h=coefh*sqrt(3*dp^2) in 3D)" />
            <cflnumber value="0.2" comment="Coefficient to multiply dt" />
        </constantsdef>
        <mkconfig boundcount="230" fluidcount="9">
            <mkorientfluid mk="0" orient="Xyz" />
        </mkconfig>
        <geometry>
            <definition dp="0.002">
                <pointref x="0" y="0" z="0" />
                <pointmin x="-4" y="0" z="-2" />
                <pointmax x="10" y="0" z="3.5" />
            </definition>
            <commands>
                <list name="GeometryForNormals">
                    <setactive drawpoints="0" drawshapes="1" />
                    <setshapemode>actual | bound</setshapemode>
                    <setmkbound mk="0" />
                    <drawextrude closed="false">
                        <extrude x="0" y="2" z="0" />
                        <point x="#-0.5-2.8-0.25" y="-1" z="#0-0.8" />
                        <point x="#4.5+2-2.8-0.25" y="-1" z="#0-0.8" />
                        <point x="10" y="-1" z="#1-0.8" />
                        <layers vdp="0.5" />
                    </drawextrude>
                    
                    <setmkbound mk="10" />
                    <drawbox>
                        <boxfill>solid</boxfill>
                        <point x="#-0.02-2.8-0.25" y="-1" z="#0-0.8" />
                        <size x="0.02" y="2" z="2" />
                        <layers vdp="0.5" />
                    </drawbox>
                    
                    <setmkbound mk="50" />
                    <drawbox>
                        <boxfill>solid</boxfill>
                        <point x="#2.8-2.8-0.25" y="-1" z="#0.7-0.8" />
                        <size x="0.5" y="2" z="0.2" />
                        <layers vdp="0.5" />
                    </drawbox>
                    <shapeout file="hdp" />
                    <resetdraw />
                </list>
                <mainlist>
                    
                    
                    <newvarcte hdp="Dp/2" />
                    <newvarcte rcyl="0.025" _rem="Radius of cylinder" />
                    <exportvar vars="hdp,rcyl" />
                    <redraw />
                    
                    
                    <setshapemode>actual | bound</setshapemode>
                    <setdrawmode mode="full" />
                    <runlist name="GeometryForNormals" times="3" />
                    
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.2" />
                        <point x="-0.2" y="1" z="-0.2" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.3" />
                        <point x="-0.2" y="1" z="-0.3" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.4" />
                        <point x="-0.2" y="1" z="-0.4" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.25" />
                        <point x="0.2" y="1" z="-0.25" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setfrdrawmode auto="true" />
                    <setmkbound mk="50" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.35" />
                        <point x="0.2" y="1" z="-0.35" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setfrdrawmode auto="true" />
                    <setmkbound mk="50" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.45" />
                        <point x="0.2" y="1" z="-0.45" />
                        <layers vdp="-0.5,-1.5,-2.5,-3.5,-4.5" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <_shapeout file="" reset="true" />
                    
                    
                    
                    
                    <runlist name="GeometryForNormals" />
                    
                    <setshapemode>actual | bound</setshapemode>
                    <setdrawmode mode="full" />
                    
                    <setmkbound mk="10" />
                    <drawbox>
                        <boxfill>solid</boxfill>
                        <point x="#-0.02-2.8-0.25" y="-1" z="#0-0.8" />
                        <size x="0.02" y="2" z="2" />
                        <layers vdp="0,-1,-2,-3" />
                    </drawbox>
                    
                    <setmkbound mk="0" />
                    <setfrdrawmode auto="true" />
                    <drawextrude closed="false">
                        <extrude x="0" y="2" z="0" />
                        <point x="#-0.5-2.8-0.25" y="-1" z="#0-0.8" />
                        <point x="#4.5+2-2.8-0.25" y="-1" z="#0-0.8" />
                        <point x="10" y="-1" z="#1-0.8" />
                        <layers vdp="0,-1,-2,-3" />
                    </drawextrude>
                    <setfrdrawmode auto="false" />
                    
                    <setmkbound mk="50" />
                    <drawbox>
                        <boxfill>solid</boxfill>
                        <point x="#2.8-2.8-0.25" y="-1" z="#0.7-0.8" />
                        <size x="0.5" y="2" z="0.2" />
                        <layers vdp="0,-1,-2,-3" />
                    </drawbox>
                    
                    
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.2" />
                        <point x="-0.2" y="1" z="-0.2" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.3" />
                        <point x="-0.2" y="1" z="-0.3" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="-0.2" y="-1" z="-0.4" />
                        <point x="-0.2" y="1" z="-0.4" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.25" />
                        <point x="0.2" y="1" z="-0.25" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.35" />
                        <point x="0.2" y="1" z="-0.35" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    <setmkbound mk="50" />
                    <setfrdrawmode auto="true" />
                    <drawcylinder radius="#rcyl">
                        <point x="0.2" y="-1" z="-0.45" />
                        <point x="0.2" y="1" z="-0.45" />
                    </drawcylinder>
                    <setfrdrawmode auto="false" />
                    
                    
                    <setmkfluid mk="0" />
                    <fillbox x="0.5" y="0" z="-0.05">
                        <modefill>void</modefill>
                        <point x="#-1-2.8-0.25" y="-1" z="#-1-0.8" />
                        <size x="20" y="2" z="1.8" />
                    </fillbox>
                    <shapeout file="" />
                </mainlist>
            </commands>
        </geometry>
        <floatings>
            <floating mkbound="50" relativeweight="0.5" />
        </floatings>
        <normals>
            <distanceh value="2.0" comment="Maximum distance (H*distanceh) to compute normals data (default=2)" />
            <geometryfile file="[CaseName]_hdp_Actual.vtk" comment="File with boundary geometry (VTK format)" />
        </normals>
        <motion>
            <objreal ref="10">
                <begin mov="1" start="0" />
                <mvnull id="1" />
            </objreal>
        </motion>
    </casedef>
    <execution>
        <special>
            <initialize>
                
                <boundnormal_cylinder mkbound="50">
                    <center1 x="-0.2" y="-1" z="-0.2" />
                    <center2 x="-0.2" y="1" z="-0.2" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                <boundnormal_cylinder mkbound="50">
                    <center1 x="-0.2" y="-1" z="-0.3" />
                    <center2 x="-0.2" y="1" z="-0.3" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                <boundnormal_cylinder mkbound="50">
                    <center1 x="-0.2" y="-1" z="-0.4" />
                    <center2 x="-0.2" y="1" z="-0.4" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                
                <boundnormal_cylinder mkbound="50">
                    <center1 x="0.2" y="-1" z="-0.25" />
                    <center2 x="0.2" y="1" z="-0.25" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                <boundnormal_cylinder mkbound="50">
                    <center1 x="0.2" y="-1" z="-0.35" />
                    <center2 x="0.2" y="1" z="-0.35" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                <boundnormal_cylinder mkbound="50">
                    <center1 x="0.2" y="-1" z="-0.45" />
                    <center2 x="0.2" y="1" z="-0.45" />
                    <radius v="#rcyl" />
                    <inside v="true" comment="Boundary particles inside the cylinder" />
                    <maxdisth v="2.0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_cylinder>
                
                <boundnormal_plane mkbound="10">
                    <point auto="true" comment="Point is calculated automatically accoding to normal configuration." />
                    <normal x="1" y="0" z="0" />
                    <maxdisth v="0" comment="Maximum distance to boundary limit. It uses H*maxdisth (default=2)" />
                </boundnormal_plane>
            </initialize>
            <damping>
                <dampingzone>
                    <limitmin x="#6.5-2.8-0.25" y="0" z="0" />
                    <limitmax x="#10-2.8-0.25" y="0" z="0" />
                    <overlimit value="1" />
                    <redumax value="10" />
                </dampingzone>
            </damping>
            <wavepaddles>
                <piston>
                    <mkbound value="10" comment="Mk-Bound of selected particles" />
                    <waveorder value="2" comment="Order wave generation 1:1st order, 2:2nd order (def=1)" />
                    <start value="0" comment="Start time (def=0)" />
                    <duration value="0" comment="Movement duration, Zero is the end of simulation (def=0)" />
                    <depth value="0.8" comment="Fluid depth (def=0)" />
                    <pistondir x="1" y="0" z="0" comment="Movement direction (def=(1,0,0))" />
                    <waveheight value="0.1" comment="Wave height" />
                    <waveperiod value="2.0" comment="Wave period" />
                    <phase value="0" comment="Initial wave phase in function of PI (def=0)" />
                    <ramp value="1" comment="Periods of ramp (def=0)" />
                    <savemotion xpos="3.0" zpos="-0.26" comment="Saves motion data. xpos and zpos are optional. zpos=-depth of the measuring point" />
                </piston>
            </wavepaddles>      
        </special>
        <parameters>
            <parameter key="SavePosDouble" value="1" comment="Saves particle position using double precision (default=0)" />
            <parameter key="Boundary" value="2" comment="Boundary method 1:DBC, 2:mDBC (default=1)" />
            <parameter key="StepAlgorithm" value="2" comment="Step Algorithm 1:Verlet, 2:Symplectic (default=1)" />
            <parameter key="VerletSteps" value="40" comment="Verlet only: Number of steps to apply Euler timestepping (default=40)" />
            <parameter key="Kernel" value="2" comment="Interaction Kernel 1:Cubic Spline, 2:Wendland (default=2)" />
            <parameter key="ViscoTreatment" value="1" comment="Viscosity formulation 1:Artificial, 2:Laminar+SPS (default=1)" />
            <parameter key="Visco" value="0.01" comment="Viscosity value" />
            <parameter key="ViscoBoundFactor" value="0" comment="Multiply viscosity value with boundary (default=1)" />
            <parameter key="DensityDT" value="3" comment="Density Diffusion Term 0:None, 1:Molteni, 2:Fourtakas, 3:Fourtakas(full) (default=0)" />
            <parameter key="DensityDTvalue" value="0.1" comment="DDT value (default=0.1)" />
            <parameter key="Shifting" value="3" comment="Shifting mode 0:None, 1:Ignore bound, 2:Ignore fixed, 3:Full (default=0)" />
            <parameter key="ShiftCoef" value="-2" comment="Coefficient for shifting computation (default=-2)" />
            <parameter key="ShiftTFS" value="1.5" comment="Threshold to detect free surface. Typically 1.5 for 2D and 2.75 for 3D (default=0)" />
            <parameter key="RigidAlgorithm" value="1" comment="Rigid Algorithm 0:collision-free, 1:SPH, 2:DEM, 3:Chrono (default=1)" />
            <parameter key="FtPause" value="0.0" comment="Time to freeze the floatings at simulation start (warmup) (default=0)" units_comment="seconds" />
            <parameter key="CoefDtMin" value="0.05" comment="Coefficient to calculate minimum time step dtmin=coefdtmin*h/speedsound (default=0.05)" />
            <parameter key="DtIni" value="0" comment="Initial time step. Use 0 to defult use (default=h/speedsound)" units_comment="seconds" />
            <parameter key="DtMin" value="0" comment="Minimum time step. Use 0 to defult use (default=coefdtmin*h/speedsound)" units_comment="seconds" />
            <parameter key="DtFixed" value="0" comment="Fixed Dt value. Use 0 to disable (default=disabled)" units_comment="seconds" />
            <parameter key="DtFixedFile" value="NONE" comment="Dt values are loaded from file. Use NONE to disable (default=disabled)" units_comment="milliseconds (ms)" />
            <parameter key="DtAllParticles" value="0" comment="Velocity of particles used to calculate DT. 1:All, 0:Only fluid/floating (default=0)" />
            <parameter key="TimeMax" value="14" comment="Time of simulation" units_comment="seconds" />
            <parameter key="TimeOut" value="0.02" comment="Time out data" units_comment="seconds" />
            <parameter key="PartsOutMax" value="1" comment="%/100 of fluid particles allowed to be excluded from domain (default=1)" units_comment="decimal" />
            <parameter key="RhopOutMin" value="700" comment="Minimum rhop valid (default=700)" units_comment="kg/m^3" />
            <parameter key="RhopOutMax" value="1300" comment="Maximum rhop valid (default=1300)" units_comment="kg/m^3" />
            <simulationdomain comment="Defines domain of simulation (default=Uses minimun and maximum position of the generated particles)">
                <posmin x="default" y="default" z="default" comment="e.g.: x=0.5, y=default-1, z=default-10%" />
                <posmax x="default" y="default" z="default" />
            </simulationdomain>
        </parameters>     
    </execution>
</case>

**Comment by chenxin202** [comment]
[testCylinder_Def.txt](https://github.com/user-attachments/files/18216798/testCylinder_Def.txt)
here  is the DEF file

**Comment by alexbexe** [comment]
We saw your email and we will try to check it as soon as possible

**Comment by chenxin202** [comment]
Thank you, Professor

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Alejandro J. C. Crespo ***@***.***>
发送时间: Monday, December 23, 2024 3:21:44 AM
收件人: DualSPHysics/DualSPHysics ***@***.***>
抄送: chenxin202 ***@***.***>; Author ***@***.***>
主题: Re: [DualSPHysics/DualSPHysics] Some errors occured in MDBC method (Discussion #285)


We saw your email and we will try to check it as soon as possible

―
